### PR TITLE
refix #2279: get associated ContestProblem with Clarification entity

### LINKED ISSUE #2279: In the contestant's clarifications block, the shortname and color of the problem are not displayed correctly.
NOTE: If this is a _discussion starter_, you need any _installation help_ or have a _question_ on how to accomplish something,
rather post at [our discussion channel](https://www.domjudge.org/chat) or send an email to our
[DOMjudge-devel mailinglist](https://www.domjudge.org/mailman/postorius/lists/domjudge-devel.domjudge.org/)
instead of filing an issue here.

### Description of the problem

image

This is the problem I set in the contest, with the correct shortname and color set.

image

This is the Clarifications page in the administrator interface, showing the correct shortname and color.

image

But this is the Clarifications page in the contestant's interface, and it doesn't show the correct shortname and color.

This should be the shortname of the problem.

### Your environment

- DOMjudge: 8.2.2
- OS: Ubuntu22.04
- webserver: nginx

### Steps to reproduce

- create a contest.
- send a clarification in contestant's interface.

On the Clarifications page in the contestant's interface, the correct shortname and color are shown.

The Clarifications page in the contestant's interface currently shows the externalid of the title and the color black.

### Any other information that you want to share?

image

My data source options are set like this.

### PR DESCRIPTION
**Pull request by moesoha**
This PR fixes #2279 indeed. In the previous commit 0219d8e173540e26b5867d219753dcb23f08fcc5, this issue was accidentally "fixed" by abusing `LEFT JOIN`. Thanks to Doctrine's mapping & collection strategies, that commit works.

This PR implements `getContestProblem(Problem $problem)` in Contest entity by moving the codes in `TwigExtension::problemBadgeForProblemAndContest`, and adds a getter for the associated ContestProblem in Clarification entity. With these helpers, the key code of #2279 (`clarification.problem.contestProblems.first`) can be replaced with `clarification.contestProblem`.

By the way, I renamed the twig filter `problemBadgeForProblemAndContest` to `problemBadgeForContest`, as the Problem is redundant here.

0219d8e173540e26b5867d219753dcb23f08fcc5 is reverted in this PR, so the bug mentioned in #2600 no longer exists.

### CODE REVIEW

File: webapp/src/Twig/TwigExtension.php [hunk 1]
```diff
@@ -1084,14 +1084,10 @@ public function problemBadge(ContestProblem $problem): string
...
+    public function problemBadgeForProblemAndContest(Problem $problem, ?Contest $contest = null): string
     {
-        foreach ($problem->getContestProblems() as $contestProblem) {
-            if ($contestProblem->getContest() === $contest) {
-                return $this->problemBadge($contestProblem);
-            }
-        }
-        return '';
+        $problem = $problem->getContestProblem($contest ?? $this->dj->getCurrentContest());
+        return $this->problemBadge($problem);
```
Comment (cubercsl) [inline webapp/src/Twig/TwigExtension.php:1090]:
```suggestion
        if ($problem === null) {
            return '';
        }
        return $this->problemBadge($problem);
```

**Comment by moesoha** [comment]
Any update? Sorry for disturbing but I would like to request a review @vmcj

**Comment by vmcj** [comment]
> Any update? Sorry for disturbing but I would like to request a review @vmcj

Hi @moesoha, I think all of us are currently very busy so it might take a bit more time.

**Review by vmcj — APPROVED** [review]
Fine by me, but I'll wait with merging for someone else to also take a look.

**Review by meisterT — APPROVED** [review]

**Pull request merged**